api/webhook/validate_auth.go

```diff
@@ -0,0 +1,170 @@
+// Copyright 2021 Chaos Mesh Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package webhook
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"reflect"
+
+	admissionv1beta1 "k8s.io/api/admission/v1beta1"
+	authv1 "k8s.io/api/authorization/v1"
+	authorizationv1 "k8s.io/client-go/kubernetes/typed/authorization/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"github.com/chaos-mesh/chaos-mesh/api/v1alpha1"
+	"github.com/chaos-mesh/chaos-mesh/pkg/selector"
+)
+
+var authLog = ctrl.Log.WithName("validate-auth")
+
+// +kubebuilder:webhook:path=/validate-auth,mutating=false,failurePolicy=fail,groups=chaos-mesh.org,resources=*,verbs=create;update,versions=v1alpha1,name=vauth.kb.io
+
+// AuthValidator validates the authority
+type AuthValidator struct {
+	Enabled bool
+	Client  client.Client
+	Reader  client.Reader
+	AuthCli *authorizationv1.AuthorizationV1Client
+
+	decoder *admission.Decoder
+
+	ClusterScoped     bool
+	TargetNamespace   string
+	AllowedNamespaces string
+	IgnoredNamespaces string
+}
+
+// AuthValidator admits a pod iff a specific annotation exists.
+func (v *AuthValidator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	if !v.Enabled {
+		return admission.Allowed("")
+	}
+
+	username := req.UserInfo.Username
+	groups := req.UserInfo.Groups
+	chaosKind := req.Kind.Kind
+
+	//var chaos v1alpha1.ChaosValidator
+
+	if chaosKind == v1alpha1.KindAwsChaos || chaosKind == v1alpha1.KindPodNetworkChaos {
+		return admission.Allowed("")
+	}
+
+	chaos := v1alpha1.GetChaosValidator(chaosKind)
+	oldChaos := v1alpha1.GetChaosValidator(chaosKind)
+	if chaos == nil {
+		err := fmt.Errorf("kind %s is not support", chaosKind)
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	err := v.decoder.Decode(req, chaos)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+	specs := chaos.GetSelectSpec()
+
+	if req.Operation == admissionv1beta1.Update {
+		// when selector is not changed, don't need to validate auth again
```

Comment:
This may cause performance problems, because we update the CR frequently, such as status. If the K8s cluster is relatively large and the user chooses a wide range, it may take a long time to select the Pod.

I think a user already has permission to create it, and is no need to verify it again if doesn't modify the scope of the experiment.

Reply:
How about only make permission test on `spec.GetSelector().Namespaces`, and do not call `selector.SelectPods` here?

If a user does not specify the namespace, actually he/she want to inject chaos into the entire-cluster, and we just check the permission about cluster-scoped(not namespace-scoped) permissions.

Reply:
In other words, if users want to restrict namespace-scoped permission, they must specify the namespace(s) that the chaos experiment will affect explicitly.

Reply:
`How about only make permission test on spec.GetSelector().Namespaces, and do not call selector.SelectPods here?`
I think it's OK. And by the way, if the namespace is not specified, the webhook will set it with the CR's namespace
https://github.com/chaos-mesh/chaos-mesh/blob/478d00d01bb0f9fb08a1085428a7da8c8f9df4e8/api/v1alpha1/common_webhook.go#L22 😅